charmcraft/commands/build.py

```diff
@@ -184,61 +205,66 @@ def build_charm(self, bases_config: BasesConfiguration) -> str:
         """
         logger.debug("Building charm in %r", str(self.buildpath))
 
-        build_env = dict(LANG="C.UTF-8", LC_ALL="C.UTF-8")
-        for key in [
-            "PATH",
-            "SNAP",
-            "SNAP_ARCH",
-            "SNAP_NAME",
-            "SNAP_VERSION",
-            "http_proxy",
-            "https_proxy",
-            "no_proxy",
-        ]:
-            if key in os.environ:
-                build_env[key] = os.environ[key]
-
-        env_flags = [f"{key}={value}" for key, value in build_env.items()]
-
-        # invoke the charm builder
-        build_cmd = [
-            "env",
-            "-i",
-            *env_flags,
-            sys.executable,
-            "-I",
-            charm_builder.__file__,
-            "--charmdir",
-            str(self.charmdir),
-            "--builddir",
-            str(self.buildpath),
-        ]
-
-        if self.entrypoint:
-            build_cmd.extend(["--entrypoint", str(self.entrypoint)])
-
-        for req in self.requirement_paths:
-            build_cmd.extend(["-r", str(req)])
-
-        retcode = polite_exec(build_cmd)
-        if retcode:
-            raise CommandError("problems running charm builder")
+        entrypoint = self.entrypoint.relative_to(self.charmdir)
+
+        # add charm files to the prime filter
+        self._set_prime_filter(entrypoint)
+
+        # set source for buiding
+        self._charm_part["source"] = str(self.charmdir)
+
+        # run the parts lifecycle
+        logger.debug("Parts definition: %s", self._parts)
+        lifecycle = PartsLifecycle(
+            self._parts,
+            work_dir=self.buildpath,
+            entrypoint=entrypoint,
+            requirements=[str(p) for p in self.requirement_paths],
+        )
+        lifecycle.run(Step.PRIME)
 
         # run linters and show the results
-        linting_results = linters.analyze(self.config, self.buildpath)
+        linting_results = linters.analyze(self.config, lifecycle.prime_dir)
         self.show_linting_results(linting_results)
 
         create_manifest(
-            self.buildpath,
+            lifecycle.prime_dir,
             self.config.project.started_at,
             bases_config,
             linting_results,
         )
 
-        zipname = self.handle_package(bases_config)
+        zipname = self.handle_package(lifecycle.prime_dir, bases_config)
         logger.info("Created '%s'.", zipname)
         return zipname
 
+    def _set_prime_filter(self, entrypoint: pathlib.Path):
+        """Add mandatory and optional charm files to the prime filter.
+
+        The prime filter should contain:
+        - The charm entry point, or the directory containing it if it's
+          not directly in the project dir.
+        - A set of mandatory charm files, including metadata.yaml, the
+          dispatcher and the hooks directory.
+        - A set of optional charm files.
+        """
+        # add entrypoint
```

Comment:
It's somewhat explained in the docstring, but I would make it explicit in both blocks in the `if`...

```
        if str(entrypoint) == entrypoint.name:
            # the entry point is in the root of the project, just include it
            self._prime.append(str(entrypoint))
        else:
            # the entry point is in a subdir, include the whole subtree starting with it
            self._prime.append(str(entrypoint.parts[0]))
```